esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -101,54 +101,151 @@ IPropertyTree * fetchESDLDefinitionFromDali(const char *esdlServiceName, unsigne
+            return NULL;
+        }
+
+        //There shouldn't be multiple entries here, but if so, we'll use the first one
+        VStringBuffer xpath("%s[@id='%s.%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+
+        return esdlBindings->getPropTree(xpath);
     }
+    catch(...)
+    {
+        DBGLOG("ESDL Binding: Unknown error encountered while fetching ESDL Binding from Dali %s[@EspProcess='%s'][@EspBinding='%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+        if (conn)
+            conn->close(false); //release lock right away
 
-    conn->close(false); //release lock right away
+    }
+    return NULL;
+}
 
-    DBGLOG("ESDL Binding: Fetching ESDL Binding from Dali %s[@EspProcess='%s'][@EspBinding='%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+IPropertyTree * fetchESDLBindingFromStateFile(const char *process, const char *bindingName, const char * stateFileName)
+{
+    Owned<IPTree> esdlState = createPTreeFromXMLFile(stateFileName);
+    if (!esdlState)
+        throw MakeStringException(-1, "Could not load DESDL binding state from local file: '%s' for ESP binding %s.%s", stateFileName, process, bindingName);
 
-    IPropertyTree *esdlBindings = conn->queryRoot();
-    if (!esdlBindings)
-       throw MakeStringException(-1, "Unable to open ESDL Service binding information in dali %s", ESDL_BINDINGS_ROOT_PATH);
+    const char * restoredBindingTS = esdlState->queryProp("@StateSaveTimems");
 
-    //There shouldn't be multiple entries here, but if so, we'll use the first one
-    VStringBuffer xpath("%s[@id='%s.%s'][1]", ESDL_BINDING_ENTRY, process, bindingName);
+    fprintf(stdout, "ESDL State restored from local state store: %s created epoch: %s", bindingName, restoredBindingTS);
 
-    return  LINK(esdlBindings->queryPropTree(xpath));
+    return esdlState->getPropTree("EsdlBinding/Binding");
+}
+
+IPropertyTree * fetchESDLBinding(const char *process, const char *bindingName, const char * stateFileName)
+{
+    Owned<IPropertyTree> esdlBinding = fetchESDLBindingFromDali(process, bindingName);
+    if (!esdlBinding)
+    {
+        return fetchESDLBindingFromStateFile(process, bindingName, stateFileName);
+    }
+
+    return esdlBinding.getLink();
+}
+
+void saveState(const char * esdlDefinition, IPropertyTree * esdlBinding, const char * stateFileFullPath)
+{
+    StringBuffer dESDLState;
+    dESDLState.setf( "<ESDLBindingState StateSaveTimems='%d'><EsdlDefinition>%s</EsdlDefinition>", msTick(), esdlDefinition ? esdlDefinition : "");
+    if (esdlBinding)
+    {
+        dESDLState.append("<EsdlBinding>");
+        toXML(esdlBinding, dESDLState);
+        dESDLState.append("</EsdlBinding>");
+    }
+
+    dESDLState.append("</ESDLBindingState>");
+
+#ifdef _DEBUG
+fprintf(stdout, "\nESDL State to be stored:\n%s", dESDLState.str());
+#endif
+
+    Owned<IPropertyTree> serviceESDLDef = createPTreeFromXMLString(dESDLState.str(), ipt_caseInsensitive);
+    saveXML(stateFileFullPath, serviceESDLDef);
```

Comment:
In case you crash while updating, I would do something like 1. write to a file with a unique name modifier. 2. rename when file is written.

If you crash and restart without dali available during #1 (could be caused by external or other thread), then you will reload from the prior cache.